xls/dslx/replace_invocations.cc

```diff
@@ -0,0 +1,504 @@
+// Copyright 2025 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xls/dslx/replace_invocations.h"
+
+#include <string>
+#include <utility>
+
+#include "absl/container/btree_set.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "xls/common/status/ret_check.h"
+#include "xls/dslx/frontend/ast.h"
+#include "xls/dslx/frontend/ast_cloner.h"
+#include "xls/dslx/frontend/ast_utils.h"
+#include "xls/dslx/frontend/module.h"
+#include "xls/dslx/import_data.h"
+#include "xls/dslx/parse_and_typecheck.h"
+#include "xls/dslx/type_system/parametric_env.h"
+#include "xls/dslx/type_system/type_info.h"
+#include "xls/dslx/type_system/typecheck_module.h"
+#include "xls/dslx/warning_collector.h"
+#include "xls/ir/bits.h"
+
+namespace xls::dslx {
+
+namespace {
+
+template <typename... T>
+struct Overloaded : T... {
+  using T::operator()...;
+};
+template <typename... T>
+Overloaded(T...) -> Overloaded<T...>;
```

Comment:
We use absl::visit and xls::Visitor for this; e.g. https://github.com/google/xls/blob/65aa30d29a36e34040707b7eaf4a8dbb5e06651d/xls/common/visitor.h#L23 you can pull in from that header